File: src/examples/cycling/CyclingRace.java
package cycling;

import gov.nasa.jpf.abstraction.Verifier;


public class CyclingRace
{
	public static void main(String[] args)
	{
		Cyclist[] cyclists = new Cyclist[3];
		
		Cyclist cl = null;
		
		// insert raw data for cyclists
		
		cl = new Cyclist();
		cl.idnum = Verifier.unknownInt(); // not relevant
		cl.time = 3725;
		cl.bonus = 5;
		cyclists[0] = cl;
		
		cl = new Cyclist();
		cl.idnum = Verifier.unknownInt(); // not relevant
		cl.time = 3569;
		cl.bonus = 10;
		cyclists[1] = cl;
		
		cl = new Cyclist();
		cl.idnum = Verifier.unknownInt(); // not relevant
		cl.time = 3766;
		cl.bonus = 50;
		cyclists[2] = cl;
		
		
		Cyclist[] results = new Cyclist[3];
		
		// compute results of the competition
        for (int i = 0; i < cyclists.length; ++i) {
			int res = cyclists[i].time - cyclists[i].bonus;

			int pos = 0;

            for (int j = i - 1; j >= 0; --j) 
			{
				if (res < results[j].time - results[j].bonus) results[j + 1] = results[j];
				else pos = j + 1;
            }

			results[pos] = cyclists[i];
		}

        int[] diffs = new int[3];
		
		Cyclist bestCL = results[0];
		int bestTime = bestCL.time - bestCL.bonus;
        diffs[0] = bestTime;
		for (int i = 1; i < results.length; ++i) {
			cl = results[i];
			int diff = Verifier.unknownInt(); // cl.time - cl.bonus - bestTime;
            diffs[i] = diff;
		}
	}

}

class Cyclist 
{
	public int idnum;
	public int time;
	public int bonus;
}

File: src/benchmarks/cycling/cyclingrace.c
#include <stdlib.h>

extern void __VERIFIER_error() __attribute__ ((__noreturn__));

void __VERIFIER_assert(int cond) {
  if (!(cond)) {
    ERROR: __VERIFIER_error();
  }
  return;
}
int __VERIFIER_nondet_int();

typedef struct {
    int idnum;
    int time;
    int bonus;
} Cyclist;

Cyclist* cyclist_new() {
    Cyclist* cl = (Cyclist*) malloc(sizeof(Cyclist));

    return cl;
}

int main() {
    Cyclist* cyclist[3];

    Cyclist* cl = NULL;

    cl = cyclist_new();
    cl->idnum = __VERIFIER_nondet_int();
    cl->time = 3725;
    cl->bonus = 5;
    cyclist[0] = cl;

    cl = cyclist_new();
    cl->idnum = __VERIFIER_nondet_int();
    cl->time = 3569;
    cl->bonus = 10;
    cyclist[1] = cl;

    cl = cyclist_new();
    cl->idnum = __VERIFIER_nondet_int();
    cl->time = 3766;
    cl->bonus = 50;
    cyclist[2] = cl;

    Cyclist* result[3];

    int i, j;

    for (i = 0; i < 3; ++i) {
        int res = cyclist[i]->time - cyclist[i]->bonus;

        int pos = 0;

        for (j = i - 1; j >= 0; --j) {
            if (res < result[j]->time - result[j]->bonus) {
                __VERIFIER_assert(j + 1 < 3);

                result[j + 1] = result[j];
            } else {
                pos = j + 1;
            }
        }

        __VERIFIER_assert(pos < 3);

        result[pos] = cyclist[i];
    }

    int diffs[3];

    Cyclist* bestcl = result[0];
    int besttime = bestcl->time - bestcl->bonus;
    diffs[0] = besttime;

    for (i = 1; i < 3; ++i) {
        cl = result[i];

        int diff = __VERIFIER_nondet_int();
        diffs[i] = diff;
    }
}
